plugins/com.google.cloud.tools.eclipse.usagetracker/src/com/google/cloud/tools/eclipse/usagetracker/AnalyticsPingManager.java

```diff
@@ -133,14 +133,30 @@ void registerOptInStatus(boolean optedIn) {
    * If the user has never seen the opt-in dialog or set the opt-in preference beforehand,
    * this method can potentially present an opt-in dialog at the top workbench level. If you
    * are calling this method inside another modal dialog, consider using {@link #sendPing(
-   * String, String, String, Shell)} and pass the {@Shell} of currently open modal dialog.
+   * String, String, String, Shell)} and pass the {@Shell} of the currently open modal dialog.
    * (Otherwise, the opt-in dialog won't be able to get input until the workbench can get input.)
+   *
+   * Safe to call from non-UI contexts.
    */
   public void sendPing(String eventName, String metadataKey, String metadataValue) {
     sendPing(eventName, metadataKey, metadataValue, null);
   }
 
-  public void sendPing(String eventName, String metadataKey, String metadataValue,
+  public void sendPing(final String eventName,
+      final String metadataKey, final String metadataValue, final Shell parentShell) {
+    if (Display.getCurrent() != null) {  // Check if we are in the UI thread.
```

Comment:
Proper check then is `display.getThread() == Thread.currentThread()`.

But do we really want to have the pings *always* happen on the UI thread?  I'd have thought we'd want the pings to always happen on a non-UI thread (e.g., via a Job), unless we have to prompt the user.